### PR TITLE
cluster-ui: database details page connected component with reusable combiner logic

### LINKED ISSUE #97885: use cluster-ui database component on CC Console
Currently we're using a separate component on CC Console. This should be replaced by the cluster-ui Database/Tables component.

Jira issue: CRDB-24945

Epic: CRDB-25476

### PR DESCRIPTION
**Pull request by THardy98**
Resolves: https://github.com/cockroachdb/cockroach/issues/97885 
Related to: https://github.com/cockroachdb/cockroach/pull/103979
Consumed by: https://github.com/cockroachlabs/managed-service/pull/13230

** DEMOS **
CC: https://www.loom.com/share/58af04f5fe1e42a6b89db7a2bd8054d0
DB: https://www.loom.com/share/ea344d07be804fb7a020d0c3d83f497f

This PR adds a connected component for the database details page on
cluster-ui to be used on CC console. Similar to the PR for the connected
databases page, we move some common selector/combiner logic between the
two connected components (for db/cc console) into cluster-ui for reuse.

We add the table details state as well to cluster-ui, making use of the
existing tableDetailsApi using sql-over-http.

Some notable changes with the existing database details page on CC console:
- We lose the delete and edit action modals, allowing users to edit/delete their database from the console
- We lose access to FGAC privileges
- We lose CC console feature flags such as `multiRegionServerlessEnabled` which are currently used to gate the `Regions` column (this is `Node/Regions` on DB console). In this PR, we instead gate the `Node/Regions` column if the user is a tenant (or when there is only 1 region).

Release note: None

### CODE REVIEW

**Comment by THardy98** [comment]
This PR is rebased off #103979 until it's merged. 
Still adding demos, struggling to get my local CC console to run properly.

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:136]: Why the concat with empty array here? Doesn't the map itself suffice?

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:133]: nit: prefer `??` over `||` when checking for nullish.

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:142]: I'm curious why we need both of these.

File: pkg/ui/workspaces/cluster-ui/src/store/databaseDetails/databaseDetails.selectors.ts
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/store/databaseDetails/databaseDetails.selectors.ts:16]: nit: IDK if this needs to be a function - wdyt?

File: pkg/ui/workspaces/db-console/src/views/databases/databaseTablePage/redux.ts
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/databases/databaseTablePage/redux.ts:74]: nit: This mapStateToProps is still using createSelector if you want to clean that up in this PR.

File: pkg/ui/workspaces/db-console/src/views/databases/indexDetailsPage/redux.ts
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/databases/indexDetailsPage/redux.ts:65]: Same with this mapStateToProps, but up to you!

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts
Comment (THardy98) [reply pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:142]: We check both fields to determine whether to refresh table details or not (i.e. if table is not loaded and not loading, we should fetch it's data).

File: pkg/ui/workspaces/db-console/src/views/databases/databaseTablePage/redux.ts
Comment (THardy98) [reply pkg/ui/workspaces/db-console/src/views/databases/databaseTablePage/redux.ts:74]: I've [refactored that](https://github.com/cockroachdb/cockroach/pull/104180/files#diff-4167077d5580805359d4e6040bb481c1fc11edb4287eb8cff040e01d07051fd4) in the follow-up table details PR

File: pkg/ui/workspaces/db-console/src/views/databases/indexDetailsPage/redux.ts
Comment (THardy98) [reply pkg/ui/workspaces/db-console/src/views/databases/indexDetailsPage/redux.ts:65]: I'll address this in the connected table details PR

File: pkg/ui/workspaces/cluster-ui/src/store/databaseDetails/databaseDetails.selectors.ts
Comment (THardy98) [reply pkg/ui/workspaces/cluster-ui/src/store/databaseDetails/databaseDetails.selectors.ts]: Agree, just using the field now.

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts
Comment (THardy98) [reply pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts]: Done.

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts
Comment (xinhaoz) [reply pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:136]: Ah got it, so we're flat mapping here. Thanks!

**Review by koorosh — APPROVED** [review]
Have some minor comments but in general :lgtm: 

> Reviewed 14 of 22 files at r2, 5 of 38 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104168#-:-NXRmNZk0JdQyk8h26-d:bbzz2lk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @maryliag, @THardy98, and @xinhaoz)

___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104168#gh-1222014543:-NXRcsYk-Q5_8Ix5L4Pd:b-bb7re0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc328e537c3d033578ae6f3ab964901c55cfbda8/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L136)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Ah got it, so we're flat mapping here. Thanks!
</blockquote></details>

til another way of doing `Array.flatMap` :)
```js
grants.flatMap(grant => grant.privileges) 
```
___
*[`pkg/ui/workspaces/cluster-ui/src/store/databaseTableDetails/tableDetails.reducer.ts` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104168#-NXRffuW2pA-pAuLc578:-NXRffuW2pA-pAuLc579:bhf3cu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f32afcaf7239c59aa943ccf1233c4e554dd8d1/pkg/ui/workspaces/cluster-ui/src/store/databaseTableDetails/tableDetails.reducer.ts#L61)):*
> ```XML
>     },
>     refresh: (state, action: PayloadAction<TableDetailsReqParams>) => {
>       state[generateTableID(action.payload.database, action.payload.table)] = {
> ```

nit. it is a bit confusing why we generate key with `generateTableID` right in reducers and for `received` and `failed` reducers we provide generated key in payload.
I would propose to make it more unified and pass in payload.
___
*[`pkg/ui/workspaces/db-console/src/views/databases/indexDetailsPage/redux.ts` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104168#-NXRmk5G63AeE_sGybGS:-NXRmk5G63AeE_sGybGT:b-24d1zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f32afcaf7239c59aa943ccf1233c4e554dd8d1/pkg/ui/workspaces/db-console/src/views/databases/indexDetailsPage/redux.ts#L68)):*
> ```XML
>     const details = stats?.data?.statistics.filter(
>       stat => stat.index_name === index, // index names must be unique for a table
>     )[0];
> ```

nit. use `find` instead of `filter` to avoid accessing array element by index
```js
    const details = stats?.data?.statistics.find(stat => stat.index_name === index)
```
___
*[`pkg/ui/workspaces/cluster-ui/src/databaseDetailsPage/helperComponents.tsx` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104168#-NXRbEHfDVtXlLBfMjDA:-NXRbEHfDVtXlLBfMjDB:b-nos8k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f32afcaf7239c59aa943ccf1233c4e554dd8d1/pkg/ui/workspaces/cluster-ui/src/databaseDetailsPage/helperComponents.tsx#L49)):*
> ```XML
>   let linkURL = "";
>   if (isCockroachCloud) {
>     linkURL = `${location.pathname}/${getMatchParamByName(
> ```

Wondered if it's safe to build the link without encoding?

**Review by THardy98 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104168)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @koorosh, @maryliag, and @xinhaoz)

___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104168#gh-1222014543:-NXVqFk208KDbrVqB2B1:b18wisy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc328e537c3d033578ae6f3ab964901c55cfbda8/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L136)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

til another way of doing `Array.flatMap` :)
```js
grants.flatMap(grant => grant.privileges) 
```
</blockquote></details>

Using this requires upgrading our `tsconfig` compiler options to `es2019` which I'm not sure if we want to do.
___
*[`pkg/ui/workspaces/cluster-ui/src/store/databaseTableDetails/tableDetails.reducer.ts` line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104168#-NXRffuW2pA-pAuLc578:-NXVqM0S4iZt75qoQ1Ml:b-fz6rrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f32afcaf7239c59aa943ccf1233c4e554dd8d1/pkg/ui/workspaces/cluster-ui/src/store/databaseTableDetails/tableDetails.reducer.ts#L61)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. it is a bit confusing why we generate key with `generateTableID` right in reducers and for `received` and `failed` reducers we provide generated key in payload.
I would propose to make it more unified and pass in payload.
</blockquote></details>

The reason for this is because the network call being done (`getTableDetails`) in the `refresh/request` sagas requires `TableDetailsReqParams`:
```
export type TableDetailsReqParams = {
  database: string;
  // Note: table name is expected in the following format: "schemaName"."tableName"
  table: string;
};
```
where we need the raw database and table name. Because of this we generate the key in their corresponding reducers.
___
*[`pkg/ui/workspaces/cluster-ui/src/databaseDetailsPage/helperComponents.tsx` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104168#-NXRbEHfDVtXlLBfMjDA:-NXVqD3O7PKZJCSXa-RP:bayr35m) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f32afcaf7239c59aa943ccf1233c4e554dd8d1/pkg/ui/workspaces/cluster-ui/src/databaseDetailsPage/helperComponents.tsx#L49)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Wondered if it's safe to build the link without encoding?
</blockquote></details>

Added `EncodeUriName`

**Review by koorosh — COMMENTED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104168)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag, @THardy98, and @xinhaoz)

___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104168#gh-1222014543:-NXVvHhWDBEpIpcZfGdX:bi6mz09) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc328e537c3d033578ae6f3ab964901c55cfbda8/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L136)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

Using this requires upgrading our `tsconfig` compiler options to `es2019` which I'm not sure if we want to do.
</blockquote></details>

agree.

**Comment by THardy98** [comment]
TYFR

**Pull request merged**
